scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -61,19 +70,24 @@ object BeamOp extends Serializable {
   )(implicit ordK: Ordering[K], kryoCoder: KryoCoder): PCollection[KV[K, java.lang.Iterable[U]]] = {
     reduceFn match {
       case ComposedMapGroup(f, g) => planMapGroup(planMapGroup(pcoll, f), g)
-      case EmptyGuard(MapValueStream(SumAll(pqm: PriorityQueueMonoid[V]))) =>
-        pcoll.apply(MapElements.via(
-          new SimpleFunction[KV[K, java.lang.Iterable[V]], KV[K, java.lang.Iterable[U]]]() {
-            override def apply(input: KV[K, lang.Iterable[V]]): KV[K, java.lang.Iterable[U]] = {
-              // We are not using plus method defined in PriorityQueueMonoid as it is mutating
-              // input Priority Queues. We create a new PQ from the individual ones.
-              // We didn't use Top PTransformation in beam as it is not needed, also
-              // we cannot access `max` defined in PQ monoid.
-              val flattenedValues = input.getValue.asScala.flatMap { value =>
-                value.asInstanceOf[PriorityQueue[V]].iterator().asScala
-              }
-              val mergedPQ = pqm.build(flattenedValues)
-              KV.of(input.getKey, Iterable(mergedPQ.asInstanceOf[U]).asJava)
+      case EmptyGuard(MapValueStream(SumAll(pqm: ScaldingPriorityQueueMonoid[v]))) =>
+        val vCollection = pcoll.asInstanceOf[PCollection[KV[K, java.lang.Iterable[PriorityQueue[v]]]]]
+
+        vCollection.apply(MapElements.via(
+          new SimpleFunction[KV[K, java.lang.Iterable[PriorityQueue[v]]], KV[K, java.lang.Iterable[U]]]() {
+            override def apply(input: KV[K, lang.Iterable[PriorityQueue[v]]]): KV[K, java.lang.Iterable[U]] = {
+
+              val topCombineFn = new TopCombineFn[v, SerializableComparator[v]](
+                pqm.count,
+                SerializableComparator[v](pqm.ordering.reverse)
+              )
+
+              @inline def flattenedValues: Stream[v] =
+                input.getValue.asScala.toStream.flatMap(_.asScala.toStream)
+
+              val outputs: java.util.List[v] = topCombineFn.apply(flattenedValues.asJava)
+              val pqs = pqm.build(outputs.asScala)
```

Comment:
why do we use both `TopCombineFn` and `pqm`? I'm confused why one isn't enough? Seems like we should be able to return `output` to me.

Reply:
`U` here is `PriorityQueue[v]` so casting to `U` fails with converting the `outputs` to a PriorityQueue.

Reply:
Ahh. Okay. It makes sense now. Can you comment?